Add new plotting module for experiment and analysis figures

conradhaupt opened the pull request:
### Summary

The following PR contains code to improve the plotting functionality of Qiskit Experiments. There are two big changes: 1) moving the visualization submodule of `curve_analysis` to its own submodule and 2) the creation of new plotter and drawer classes which managed plotting figures and interfacing with plotting backends respectively.

### Details and Comments

The original `BaseCurveDrawer` class from `qiskit_experiments.curve_analysis.visualization` was used by `CurveAnalysis` to draw scatter points, curve-fits, and fit-reports. However, this limited usage of `BaseCurveDrawer` in other parts of Qiskit Experiments and for figures other than a curve-fit, such as IQ plots.

The new module (`qiskit_experiments.visualization`) contains two new submodules `plotters` and `drawers`, which are the core of this PR. All other code from `qiskit_experiments.curve_analysis.visualization`, that hasn't been modified much, has also been moved to the new submodule. Plotters and drawers fulfill different functions, such that analysis classes interact with a plotter which intern interacts with a drawer. A drawer implements a generic interface (`BasePlotter`) for a plotting backend, such as Matplotlib. A plotter deals with managing data to be plotted and invoking specific drawing functions of a drawer. `CurveAnalysis` now contains a plotter instead of a drawer.

![image](https://user-images.githubusercontent.com/6696704/190595412-90e5708c-0f82-45b5-9ed8-e93a2303e5d1.png)

`BasePlotter` implements an interface to set plotting data and generate a figure. Subclasses of `BasePlotter` override `_plot_figure()`, where drawing methods are called on `self.drawer` (i.e, `drawer.draw_*`). Implemented subclasses of `BasePlotter` and `BaseDrawer` are `CurvePlotter` and `MplDrawer` respectively. You can refer to `MplDrawer` and `CurvePlotter` for example usage and implementations of these base-classes. There are three important stages to plotting with this new module:

1. Setting data for plotting (done by an analysis class or user).
2. Setting plotter and drawer options.
3. Generating and returning the figure (for the plotter, this involves calling drawer `draw_*` functions).

#### Setting data for plotting

Plotters separate data into two groups: **series** and **figure** data. Series data is a nested dictionary where the first key is a series (or curve) name (called `series_name`) and the second key (called `data_key`) identifies the data associated with the series. Figure data is also a dictionary, but only of depth-one, also indexed by a `data_key`. Series data contains data-values to be plotted that are related to a given series, curve, or fit model. A series can be seen as a single entry in a legend. Figure data contains data-values or information related to the entire plot/figure; such as a fit report or text to be drawn on the canvas. Helper functions exist in `BasePlotter` to assist subclasses with querying and setting series and figure data: `data_for`, `data_exists_for`, `data_keys_for`, `set_series_data`, `set_figure_data`. A plotter subclass defines the data-keys that are accepted in `expected_series_data_keys()` and `expected_figure_data_keys()`, but only to raise a warning to users when `set_series_data` and `set_figure_data` are called. Unexpected data-keys are still added to the internal data storage.

#### Setting options

Both `BasePlotter` and `BaseDrawer` have two sets of options: normal options (`<plotter/drawer>.options`) and figure-options (`<plotter/drawer>.figure_options`). Normal options, called drawer and plotter options in the two classes, contain parameters that control **how** the classes function. Examples include the subplots tuple which defines the number of sub-axes to use and the initial axis object. For `CurvePlotter`, an additional option is `plot_sigma` which is used when drawing confidence intervals. Figure-options are options that define **what** is plotted/drawn on the canvas. This includes the axis labels, axis units, and figure title. The distinction goes further in that the first step taken, when a plotter generates a figure, is to pass its figure-options to `plotter.drawer`. The figure-options passed to the drawer are the intersection between `plotter.figure_options` and `plotter.drawer.figure_options`, where the plotter has priority. For example,

```python
plotter.set_figure_options(x=0, y=1)
plotter.drawer.set_figure_options(x="a", z="b")
# After plotter.figure() called.
opts = plotter.drawer.figure_options
opts.x	# returns 0, not "a"
opts.y	# returns 1
opts.z	# raise an exception as z doesn't exist in drawer.figure_options
```

By separating options from figure-options, `BasePlotter` can control which options overwrite their counterpart in `plotter.drawer`. A user of `BasePlotter` and its subclasses is aware (after reading the documentation) that `plotter.figure_options` contains information that is drawn onto the canvas. Advanced users would be aware that `drawer.figure_options` is updated by the plotter. By keeping the two separate --- and updating the drawer when generating a figure --- a user, or class isn't forced to interact with the nested drawer instance.
As well as splitting options into two sets, this PR introduces a new style class `PlotStyle` to contain style parameters such as axis-label text size and the default figure size. There was a legacy class `PlotterStyle` used by `fit_result_plotters.py`, though the code didn't appear to be used anywhere in Qiskit Experiments. `PlotStyle` takes inspiration from `PlotterStyle` but instead subclasses `Options`, with new methods to merge and update instances. `BaseDrawer` has two style instances, a default style (an option) and custom style (a figure-option), where the full style for a drawer instance is their combined style (see `BaseDrawer.style`). `BasePlotter` only has one style instance (an option) as the drawing backend is more likely to define extra style parameters. `plotter.drawer.figure_options.custom_style` is assigned the value of `plotter.options.style` when `plotter.figure()` is called, and thus a user does not need to interact with it.

#### Generating and returning the figure

Subclasses of `BasePlotter` represent a type of figure, such as a curve-fit (`CurvePlotter`) or IQ plot. The core functions to be overridden in such a subclass are `_default_options`, `_default_figure_options`, and `_plot_figure`. The last of the three functions is called by `BasePlotter.figure()`, and is where calls to `plotter.drawer.draw_*` should be made. `BaseDrawer` contains the same, or equivalent, drawing functions as `BaseCurveDrawer`, from before this PR.

Given that the figure is only generated when accessed (i.e., when calling `plotter.figure()`), we delay drawing onto the canvas until all relevant data has been loaded into the plotter, via the data assignment functions. The original procedure for generating a figure using `BaseCurveDrawer` was to initialize the canvas, draw, then format the canvas. This procedure is kept, but is now contained in `BasePlotter.figure()`. The object returned is the same object returned by `plotter.drawer.figure`.

#### Tests

- I have implemented end-to-end tests for `BasePlotter`, `BaseDrawer`, and `MplDrawer`. There are also tests for the new `PlotStyle` class and whether `plotter.options` and `plotter.figure_options` are passed to `plotter.drawer` correctly.
- I have verified that the following experiments still generate the same figures with the new visualization module. Note that there was a bug with the legend-location style option before this PR, which is now fixed. This is why legends may be located differently.
### `CrossResonanceHamiltonian`
	
![image](https://user-images.githubusercontent.com/6696704/190595488-0cdbaf7f-1212-40d6-a2b7-bfd6d98075e9.png)

### `T2Ramsey`

![image](https://user-images.githubusercontent.com/6696704/190595504-73e5aece-7ee5-48f6-8944-239cdc5ed10f.png)


### `QubitSpectroscopy`

![image](https://user-images.githubusercontent.com/6696704/190595523-b00ae650-f827-40a8-a9c5-4b59513f2a41.png)


### `StandardRB`

![image](https://user-images.githubusercontent.com/6696704/190595538-1efb99ef-7a2a-4946-9695-d9465079331c.png)

## Questions for Reviewers

- ~In refactoring and moving `qiskit_experiments.curve_analysis.visualization`, I noticed that `fit_result_plotters.py` was not used anywhere in QE. Should it be removed?~
- I am not 100% sure how to ensure options and plot-options are added to the generated documentation. It looks like there are custom styles for experiment and analysis classes, but it may be overboard to do this for the plotter and drawer functionality.

## Still To Do

- [x] ~~Complete serialization of `BasePlotter` and `BaseDrawer`.~~ (implemented in c40594c)
	- plotter data is not currently serialized. I don't think it should be as it isn't needed after figures have been generated and the data exists in the experiment data anyway.
- [x] Add release notes.

conradhaupt commented:
@eggerdj and @nkanazawa1989, it would be great if you could review this PR. Let me know if you have any questions about the implementation.

eggerdj commented on qiskit_experiments/curve_analysis/base_curve_analysis.py:
```diff
@@ -113,16 +113,16 @@ def models(self) -> List[lmfit.Model]:
-        """A short-cut for curve drawer instance."""
-        return self._options.curve_drawer
+    def plotter(self) -> BasePlotter:
+        """A short-cut for curve plotter instance."""
```
```suggestion
        """A short-cut to the curve plotter instance."""
```

conradhaupt replied:
Addressed in 8c588dc.

eggerdj commented on qiskit_experiments/curve_analysis/composite_curve_analysis.py:
```diff
@@ -124,9 +124,9 @@ def models(self) -> Dict[str, List[lmfit.Model]]:
-        """A short-cut for curve drawer instance."""
-        return self._options.curve_drawer
+    def plotter(self) -> BasePlotter:
+        """A short-cut for plotter instance."""
```
```suggestion
        """A short-cut to the plotter instance."""
```

eggerdj commented on qiskit_experiments/curve_analysis/composite_curve_analysis.py:
```diff
@@ -251,22 +247,22 @@ def _run_analysis(
-                        x_data=sub_data.x,
-                        y_data=sub_data.y,
-                        name=model._name + f"_{analysis.name}",
+                    self.plotter.set_series_data(
```
Should this be called `add_series_data`?

conradhaupt replied:
I did think about using `add_series_data` and `add_figure_data` but chose `set_*` as repeated calls to `set_*_data` with the same data-keys will overwrite the associated data in the plotter. `add_*` would imply that there is a collection of entries per series-name and data-key, and that the data is not overwritten; which didn't match up with prior usage of `BaseCurveDrawer`. The original drawer class and the analysis classes that use it identify data as single NumPy array variables (`x`, `y_err`, `y_mean`), per series-/model-name, and not lists of NumPy arrays. If we want `BasePlotter` to accept multiple NumPy arrays per data-key without overwriting previous values, then I do think the functions should be renamed to `add_*_data`. Though in this case we would need to define how a subclass should treat multiple data entries for the same data-key.

eggerdj replied:
Ok in this case `set` makes sense.

eggerdj commented on qiskit_experiments/visualization/__init__.py:
```diff
@@ -0,0 +1,90 @@
+
+.. currentmodule:: qiskit_experiments.visualization
+
+Visualization provides plotting functionality for creating figures from experiment and analysis results.
+This includes plotter and drawer classes to plot data in :py:class:`CurveAnalysis` and its subclasses.
```
Add a bit more detail here. Mention the bridge design pattern and e.g. mention that this enables different drawer backends such as mpl and plotly (even though we do not implement plotly)

eggerdj commented on qiskit_experiments/visualization/drawers/base_drawer.py:
```diff
@@ -0,0 +1,330 @@
+
+    format_canvas
+
+        This method should implement a protocol to format the appearance of canvas. Typically, it updates
```
```suggestion
        This method formats the appearance of the canvas. Typically, it updates
```

eggerdj reviewed (commented):
Here are a few points on the first few files.

eggerdj commented on qiskit_experiments/visualization/fit_result_plotters.py:
```diff
@@ -21,19 +21,56 @@
+class PlotterStyle:
+    """A stylesheet specific for :mod:`fit_result_plotters`.
+
+    This style class is used by :mod:`fit_result_plotters`, but not by :class:`BasePlotter` or
```
This first sentence seems circular. `PlotterStyle` is for `fit_result_plotters` but it is defined in `fit_result_plotters`. Which part of `fit_result_plotters` is using this? Can you clarify in the docstring?

conradhaupt replied:
Removed file in f6b6774.

eggerdj commented on qiskit_experiments/visualization/plotters/base_plotter.py:
```diff
@@ -0,0 +1,453 @@
+    instead only associated with the figure. Examples include analysis reports or other text that is
+    drawn onto the figure canvas.
+
+    # TODO: Add example usage and description of options and plot-options.
```
Yes please add this.

conradhaupt replied:
Added in 8c588dc.

eggerdj commented on qiskit_experiments/visualization/plotters/base_plotter.py:
```diff
@@ -0,0 +1,453 @@
+        # Data for figure-wide drawing, unrelated to series data, such as text or fit reports.
+        self._figure_data: Dict[str, Any] = {}
+
+        # Options for the plotter
+        self._options = self._default_options()
+        # Plotter options that have changed, for serialization.
+        self._set_options = set()
+
+        # Plot options that are updated in the drawer when `plotter.figure()` is called
+        self._plot_options = self._default_plot_options()
+        # Plot options that have changed, for serialization.
+        self._set_plot_options = set()
```
Can you add a comment, either here or in the class description above, why we have `_options` and `_plot_options`? Something to explain when an option should be in one or the other variable? The current comments hint at this but need to be more explicit.

conradhaupt replied:
Addressed in 8c588dc.

eggerdj commented on qiskit_experiments/visualization/plotters/base_plotter.py:
```diff
@@ -0,0 +1,453 @@
+        # Plot options that have changed, for serialization.
+        self._set_plot_options = set()
+
+        # The drawer backend to use for plotting.
+        self._drawer = drawer
+
+    @property
+    def drawer(self) -> BaseDrawer:
+        """The drawer being used by the plotter."""
+        return self._drawer
+
+    @drawer.setter
+    def drawer(self, new_drawer: BaseDrawer):
+        """Set the drawer to be used by the plotter."""
+        self._drawer = new_drawer
```
If we can set and get this would it not make more sense to simply have `self.drawer = drawer` and remove the properties?

conradhaupt replied:
My intention with this was to allow `BasePlotter` subclasses to execute code when `drawer` was changed. However, this is not used in this PR and I don't foresee it being used by any subclasses in the immediate future. I'll replace this with just a public `.drawer` variable.

conradhaupt replied:
Addressed in 8035121.

eggerdj commented on qiskit_experiments/visualization/plotters/base_plotter.py:
```diff
@@ -0,0 +1,453 @@
+
+        Series data includes data such as scatter points, interpolated fit values, and
+        standard-deviations. Series data is grouped by series-name and then by a data-key, both strings.
+        Though series data can be accessed through :meth:`series_data`, it is recommended to use
+        :meth:`data_for` and :meth:`data_exists_for`.
```
```suggestion
        Though series data can be accessed through :meth:`series_data`, it is recommended to access
        them with :meth:`data_for` and :meth:`data_exists_for`.
```
why by the way? Can you add a rational for this?

eggerdj commented on qiskit_experiments/visualization/plotters/base_plotter.py:
```diff
@@ -0,0 +1,453 @@
+        """Returns a list of data-keys for the given series.
+
+        Args:
+            series_name: The series name for the given series.
```
This reads in a circular fashion.
```suggestion
            series_name: The series name for which to return the data-keys, i.e., the types of data for each series.
```

eggerdj commented on qiskit_experiments/visualization/plotters/base_plotter.py:
```diff
@@ -0,0 +1,453 @@
+            list: The list of data-keys for data in the plotter associated with the given series. If the
+                series has not been added to the plotter, an empty list is returned.
+        """
+        if series_name not in self._series_data:
+            return []
+        return list(self._series_data[series_name])
```
```suggestion
        return list(self._series_data.get(series_name, [])
```

conradhaupt replied:
Addressed in 8035121.

eggerdj commented on qiskit_experiments/visualization/plotters/base_plotter.py:
```diff
@@ -0,0 +1,453 @@
+            x, y, yerr = data.x, data.y, data.yerr
+
+        :meth:`data_for` is intended to be used by sub-classes of :class:`BasePlotter` when
+        plotting in :meth:`_plot_figure`.
```
```suggestion
        plotting in the :meth:`_plot_figure` method.
```

conradhaupt replied:
Addressed in 7732065.

eggerdj commented on qiskit_experiments/visualization/plotters/base_plotter.py:
```diff
@@ -0,0 +1,453 @@
+        return self._options
+
+    @property
+    def plot_options(self) -> Options:
```
Give the description, should we rename this to `drawer_options` or `options_for_drawers`? Same for the associated internal variable?

conradhaupt replied:
@nkanazawa1989 also brought this up in another comment. I've renamed `plot_options` to `figure_options` based on his suggestion. Looking back at the code, it would be good if users of `BasePlotter` were oblivious to drawers, unless otherwise necessary. See 956696a for the changes.

eggerdj commented on qiskit_experiments/visualization/plotters/base_plotter.py:
```diff
@@ -0,0 +1,453 @@
+        self.drawer.set_plot_options(custom_style=self.options.style)
+
+        ## Plot Options
+        # HACK: We need to accesses internal variables of the Options class, which is not good practice.
+        # Options._fields are dictionaries. However, given we are accessing an internal variable, this
+        # may change in the future.
```
This is not super nice but I also cannot think of a better way to do this given the way that `Options` is built.

conradhaupt replied:
I've updated `PlotStyle` to have a `__dict__` property. There was some confusion I had with how `Options` defines `__dict__` and `_fields`, given that it has some Python magic to achieve the functionality it requires. Looking back at its implementation allowed me to properly implement a `__dict__` property for `PlotStyle. See 9f4e1ee for the changes and additional tests.

eggerdj commented on qiskit_experiments/visualization/plotters/curve_plotter.py:
```diff
@@ -0,0 +1,119 @@
+                showing the configuration to write confidence intervals for the fit curve.
+                The first argument is the relative sigma (n_sigma), and the second argument is
+                the transparency of the interval plot in ``[0, 1]``.
+                Multiple n_sigma intervals can be drawn for the single curve.
```
```suggestion
                Multiple n_sigma intervals can be drawn for the same curve.
```

eggerdj reviewed (commented):
This looks overall good to me. The structure looks fine. I have a few minor nitpicks.

nkanazawa1989 reviewed (commented):
Thanks @conradhaupt , this is really great start of designing a unified visualization framework for QE. Especially I like the separation of plotter and drawer, which allows us to easily switch visualization backend. Indeed, not only backend itself, but also we can modify the appearance output image from the same backend by providing user-defined drawer. 

My concern is mainly about flexibility. For example, if we want to draw extra `axhline` (in terms of matplotlib), we need to define custom data structure in the plotter and write new drawing method for drawer. Namely such analysis class should define both plotter and drawer subclass. This is because current methods are designed based off of the context of visualization, while other popular visualization packages define data structure based on abstraction of visualized object. I think such pattern is not always perfect, but worth revisiting the data structure so that an experiment author can prepare plotter instance for non-curve analysis type visualization (or some extended one) with minimum overhead.

Apart from flexibility, this PR actually includes several breaking API changes. Most critical one is the module path since the experiment Json loader identifies the class by import path. In this sense, I would leave conventional drawer as-is with addition of deprecation warning, and replaces the default plotter in existing analysis classes with new one (this also prevents breaking the user code). I think it is also necessary to write upgrade release note which would include a mini-migration guide for developers.

conradhaupt commented on qiskit_experiments/curve_analysis/base_curve_analysis.py:
```diff
@@ -113,16 +113,16 @@ def models(self) -> List[lmfit.Model]:
-    def drawer(self) -> BaseCurveDrawer:
-        """A short-cut for curve drawer instance."""
-        return self._options.curve_drawer
+    def plotter(self) -> BasePlotter:
```
Plotters and drawers are different in that you plot a figure but draw lines or shapes on a canvas. Given that this PR uses plotters as a bridge/intermediary between analyses and drawing on a canvas, it makes sense that the analysis class contains the plotter, which intern contains the drawer. With the new classes, and their names, the analysis classes need to access the plotter instead of the drawer, hence the name-change. If `.drawer` needs to be kept for deprecation purposes, I can make it return `.plotter.drawer` until it's removed in the future.

conradhaupt replied:
Partially addressed in 7101687. I intend to add a new `BaseDrawer` subclass to wrap deprecated `BaseCurveDrawer` instances. This should allow users to run old code during the deprecation period.

conradhaupt replied:
Addressed fully in 28a8c80.

conradhaupt commented on qiskit_experiments/library/quantum_volume/qv_analysis.py:
```diff
@@ -20,7 +20,7 @@
 import uncertainties
 from qiskit_experiments.exceptions import AnalysisError
-from qiskit_experiments.curve_analysis import plot_scatter, plot_errorbar
+from qiskit_experiments.visualization import plot_scatter, plot_errorbar
```
Correct, this analysis class has not been migrated to the new framework yet. The goal is to migrate this to use a plotter, but most likely this will happen in a follow-up PR.

conradhaupt commented on qiskit_experiments/visualization/drawers/base_drawer.py:
```diff
@@ -0,0 +1,330 @@
+                displayed in the scientific notation.
+            yval_unit (str): Unit of y values. See ``xval_unit`` for details.
+            figure_title (str): Title of the figure. Defaults to None, i.e. nothing is shown.
+            series_params (Dict[str, Dict[str, Any]]): A dictionary of plot parameters for each series.
```
I think `series_param` should be part of `BaseDrawer` as it defines backend-/drawer-independent properties of different series. If it was a subclass option, then `BasePlotter` would need a different way of defining series-specific properties, such as their colour and axis/canvas index, which would then be passed to different drawer subclasses in subclass specific ways.

conradhaupt commented on qiskit_experiments/visualization/drawers/base_drawer.py:
```diff
@@ -0,0 +1,330 @@
+        """Final cleanup for the canvas appearance."""
+
+    @abstractmethod
+    def draw_raw_data(
```
I completely agree with renaming the drawer functions to be more generic. I will go ahead and make these changes.

Regarding implementing QVPlotter or not: one of the goals of this framework is that a BasePlotter subclass defines a kind of figure being plotted. This can be a curve-fit/X-Y plot or an IQ level-1 plot. Obviously, this can be changed slightly so that maybe a given subclass is applicable to CurveAnalysis and QV analysis. I can see that there may be additional drawer functions that would allow the module to recreate figure3a, but I'll leave those additional functions for a follow-up PR.

conradhaupt replied:
Addressed `BaseDrawer` function names in 94158f4.

conradhaupt commented on qiskit_experiments/visualization/fit_result_plotters.py:
```diff
@@ -21,19 +21,56 @@
+    )
 
 
 class MplDrawSingleCanvas:
```
Addressed in f6b6774, 492fdce, and f9d5ab4.

conradhaupt commented on qiskit_experiments/visualization/plotters/base_plotter.py:
```diff
@@ -0,0 +1,453 @@
+            drawer: The drawer to use when creating the figure.
+        """
+        # Data to be plotted, such as scatter points, interpolated fits, and confidence intervals
+        self._series_data: Dict[str, Dict[str, Any]] = {}
```
Series-data would also be used for QV, even if there is only one series name or the data-value is a single float (like a threshold value). Maybe there's a better name for `figure_data` to avoid that confusion?

I'm opposed to having abstract drawing shapes as the data-storage for `BasePlotter`, as this would further mix plotting code with analysis classes. Plotter subclasses are meant to be a bridge between analyses and the figure canvas, separating their responsibilities. However, I can definitely see a scenario a data-frame is used instead, as is mentioned in #893. A data-frame would have labels (data-keys) and indices (series-names) for values, which is all `BasePlotter` needs.

conradhaupt replied:
I like the idea of renaming `figure_data` to `supplementary_data`. It makes sense as the core data being plotted is contained in `series_data`.

conradhaupt replied:
Addressed in f1d1fc2.

conradhaupt commented on qiskit_experiments/visualization/style.py:
```diff
@@ -0,0 +1,118 @@
+from qiskit_experiments.framework import Options
+
+
+class PlotStyle(Options):
```
Each `BaseDrawer` subclass must define a default style, which should inherit from the parent drawer. This allows users to only define style-parameters they want to overwrite. The default style of a drawer subclass is defined in `_default_style()`, allowing subclasses to add extra style parameters easily. This is a similar mechanism to `options`, `plot_options`, and `experiment_options`. To be consistent with other `_default_*` functions, the style type would have to have an interface similar to `Options`. To combine the default style with the custom style, defined by a user, the style type also needs to allow merging and updating; exactly like a dictionary. It was easier to extend `Options` --- and add `merge()` and `update()` --- than subclass `dict` and add functionality equivalent to that provided by `Options`. This inheritance also means `PlotStyle` benefits from value validation.

conradhaupt replied:
I completely agree that nested `Options` are actually bad. I checked, and `Options` doesn't support "namespaced" options being set unless it's as a dictionary key (`opts["key-name"] = ...`), which limits setting them using ``plotter.options.style.foo1 = "boo"``. Having a dedicated class --- instead of merging styles with `options` --- also allows `BaseDrawer` to separate the default style parameters from custom style parameters, so that users can view only those options which have changed. This is the reason I added `default_style` and `custom_style` options. The same functionality could be achieved using`options` and a set of altered options, but then it wouldn't be clear which options are style parameters and which aren't as the names may not be clear; unless we use namespaced names (which is difficult with `Options`).

Given other comments also bring into question whether `PlotStyle` should subclass `Options`, I have refactored the `PlotStyle` class to inherit from `dict` (see 7c3ca90 and 8ed2c9f). I have kept style parameters separate from `options` and `figure_options` be keeping the `default_style` and `custom_style` options. This allows `BaseDrawer` subclasses to merge and update style parameters (see `BaseDrawer.style`) while keeping them in a manageable container. Regarding namespacing style parameters, the fullstop in variable names makes it difficult to use PlotStyle. For example, style parameters cannot be set in `__init__` as kwargs (i.e., `PlotStyle(tick.text_size=14)` doesn't work). Because of this, I've left the style parameter names with underscores instead of fullstops. I do think the namespaced/grouped interface would be useful, and should be considered as the visualization module is extended.

conradhaupt commented on qiskit_experiments/curve_analysis/base_curve_analysis.py:
```diff
@@ -168,7 +168,7 @@ def _default_options(cls) -> Options:
         options = super()._default_options()
 
-        options.curve_drawer = MplCurveDrawer()
+        options.plotter = CurvePlotter(MplDrawer())
```
Partially addressed in 7101687.

conradhaupt replied:
Addressed fully in 28a8c80.

conradhaupt commented on qiskit_experiments/visualization/plotters/base_plotter.py:
```diff
@@ -0,0 +1,453 @@
+            series_params={},
+        )
+
+    def set_options(self, **fields):
```
That's a very good point, I forgot about options validation. I've addressed this in 9f4e1ee. I left out validation of `BasePlotter.plot_options` as these are passed to `plotter.drawer` and thus depend on which options are set there. Plot options are copied across and validated in `plotter.figure()` (`_initialize_drawer()`).

conradhaupt commented on qiskit_experiments/visualization/plotters/base_plotter.py:
```diff
@@ -0,0 +1,453 @@
+        """Returns the expected figures data-keys supported by this plotter."""
+
+    @property
+    def options(self) -> Options:
```
I completely understand that. @eggerdj suggested `drawer_options` or `options_for_drawer`, which are also good. In hindsight, it would be good for users to be oblivious of the drawer unless otherwise necessary. If that's the case, maybe the options should be renamed to `options` and `figure_options`; where `figure_options` are options/parameters for the figure itself (like axis labels etc.) whereas `options` are parameters that control the behaviour of the plotter or drawer. I've made this change in 956696a, but we can continue discussing it here.

conradhaupt commented:
@nkanazawa1989 and @eggerdj. I think I've addressed most of your feedback from the review. I've made the following changes:

- Refactor `BaseDrawer` to be more generic instead of tailored to `CurveAnalysis`.
- Re-added `.curve_analysis.visualization` module for deprecation purposes.
- Added new `LegacyCurveCompatDrawer` to support deprecated `BaseCurveDrawer` until removed.
	- Used by `BaseCurveAnalysis` and `CompositeCurveAnalysis`.
- Fix, expand, and edit docstrings.
- Rename `plot_options` in `BasePlotter` and `BaseDrawer` to `figure_options`.
- Add section in docs on `options` and `figure_options`.
- Add checks for invalid fields for `options` and `figure_options`.
- Fix tests failing for raised `DeprecationWarning` from `.curve_analysis.visualization` module.
	- Needed as not all analysis classes have been moved over to the new visualization module.

#### `LegacyCurveCompatDrawer`

To demonstrate the difference in functionality between the new module and the compatibility drawer `LegacyCurveCompatDrawer`, I have generated an arbitrary figure as an example. The left column contains axes drawn by the new `CurvePlotter` and `MplDrawer` whereas the right column contains axes drawn by the old `MplDrawer` (subclass of `BaseCurveDrawer`) wrapped in `LegacyCurveCompatDrawer`.

![image](https://user-images.githubusercontent.com/6696704/192509433-3be29acb-edff-479e-b0c2-fe16869f4cb7.png)

This compatibility class is already deprecated, so it can be removed alongside `qiskit_experiments.curve_analysis.visualization`. Note that the legends are located in different parts of the axes. This is because of a bug in the original `BaseCurveDrawer` which is fixed in the new `BaseDrawer`.

conradhaupt commented:
> @nkanazawa1989 and @eggerdj. I think I've addressed most of your feedback from the review. I've made the following changes:

For some reason, I cannot re-request review from both of you as GitHub removes the other request when adding the second. Sorry for the spam.

nkanazawa1989 commented on qiskit_experiments/visualization/plotters/curve_plotter.py:
```diff
@@ -0,0 +1,137 @@
+            "y_formatted",
+            "y_formatted_err",
+            "x_interp",
+            "y_mean",
+            "sigmas",
```
```suggestion
            "y_interp",
            "y_interp_err",
```
to align with `_formatted`. I think changing names are okay because the logic to generate figure is now accessed only internally.

conradhaupt replied:
Addressed in 9a909f5 and fdf1c23.

nkanazawa1989 reviewed (commented):
Thanks @conradhaupt . The updates in the previous commits are pretty nice. I'm happy with new classes. The relation ship between plotter and drawer is now much clear. Basically, we
- subclass plotter depending on the purpose of figure
- subclass drawer for different drawing backend

and new drawer methods are sufficiently agnostic to the plotter. I added some more comments mainly about option hierarchy.

eggerdj commented on qiskit_experiments/visualization/plotters/base_plotter.py:
```diff
@@ -0,0 +1,517 @@
+        self._figure_options.update_options(**fields)
+        self._set_figure_options = self._set_figure_options.union(fields)
+
+    def _initialize_drawer(self):
```
The name in the doc string says it all. How about `def _configure_drawer` or maybe even better `def _prepare_drawer()`?

conradhaupt replied:
Addressed in 84ef573.

eggerdj reviewed (commented):
I only have minor comments. Overall this looks good.

conradhaupt commented on qiskit_experiments/visualization/drawers/mpl_drawer.py:
```diff
@@ -0,0 +1,456 @@
+        ind = self._series.index(name) % len(self.DefaultMarkers)
+        return self.DefaultMarkers[ind]
+
+    def _update_label_in_dict(
```
I agree. Addressed in b604350.

conradhaupt commented on qiskit_experiments/library/characterization/analysis/cr_hamiltonian_analysis.py:
```diff
@@ -60,20 +61,24 @@ def __init__(self):
-        default_options.curve_drawer.set_options(
+        default_options.plotter.set_options(
             subplots=(3, 1),
+            style=PlotStyle(
```
Wrapping style parameters in a class, or another container, helps to group them for later processing and portability. Subclasses of `BaseDrawer` need to be able to 1) set default style parameters which are subclass compatible and 2) have plotters or users override certain style parameters. This could be accomplished as standard options, but then it would be confusing as to if an option is a style parameter or drawer option that modifies behaviour. Matplotlib uses rcparams, [Qiskit Terra](https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/visualization/pulse_v2/stylesheet.py), and the old `curve_analysis.visualization` module uses a dedicated class. However, I agree with your other comment that `PlotStyle` should rather be a dictionary to avoid nested `Options` instances.

conradhaupt replied:
I've converted PlotStyle to inherit from `dict` instead of `Options` in 7c3ca90.

conradhaupt commented on qiskit_experiments/curve_analysis/__init__.py:
```diff
@@ -318,8 +318,8 @@ class AnalysisB(CurveAnalysis):
 Afterwards, the analysis draws several curves in the Matplotlib figure.
-User can set custom drawer to the option ``curve_drawer``.
-The drawer defaults to the :class:`MplCurveDrawer`.
+User can set custom plotter to the option ``plotter``.
```
Addressed in 6bdc6ef.

conradhaupt commented on qiskit_experiments/curve_analysis/composite_curve_analysis.py:
```diff
@@ -310,23 +353,16 @@ def _run_analysis(
                         )
                         y_mean = unp.nominal_values(y_data_with_uncertainty)
                         # Draw fit line
```
Addressed in 6bdc6ef.

conradhaupt commented on qiskit_experiments/visualization/__init__.py:
```diff
@@ -0,0 +1,73 @@
+
+    BaseDrawer
+    MplDrawer
+    LegacyCurveCompatDrawer
```
Understood. It is only used for backwards compatibility. Addressed in 6bdc6ef.

conradhaupt commented on qiskit_experiments/visualization/__init__.py:
```diff
@@ -0,0 +1,73 @@
+
+# PlotStyle is imported by .drawers and .plotters. Skip PlotStyle import for isort to prevent circular
+# import.
+from .style import PlotStyle  # isort:skip
```
VSCode uses Pylints `[isort](https://pycqa.github.io/isort/)` program to sort imports, which also sorts alphabetically. This comment stopped `iosrt` from sorting this import, to prevent a circular import. However, I've managed to fix the circular import and thus removed the comment in 7c3ca90.

conradhaupt commented on qiskit_experiments/visualization/drawers/base_drawer.py:
```diff
@@ -0,0 +1,387 @@
+    draw_line
+
+        This method plots a line from provided X and Y values. This method is typically called with
+        interpolated x and y values from a curve-fit.
```
Addressed in 6bdc6ef.

conradhaupt commented on qiskit_experiments/visualization/drawers/base_drawer.py:
```diff
@@ -0,0 +1,387 @@
+    draw_text_box
+
+        This method draws a text-box on the canvas, which is a rectangular region containing some text.
+        This method is typically called with a list of analysis results and reduced chi-squared values
```
Addressed in 6bdc6ef.

conradhaupt commented on qiskit_experiments/visualization/drawers/base_drawer.py:
```diff
@@ -0,0 +1,387 @@
+    ==========================
+
+    Drawers have both :attr:`options` and :attr:`figure_options` available to set parameters that define
+    how to drawer and what is drawn. :class:`BasePlotter` is similar in that it also has ``options`` and
```
Addressed in 6bdc6ef.

conradhaupt commented on qiskit_experiments/visualization/drawers/base_drawer.py:
```diff
@@ -0,0 +1,387 @@
+        legend_label: Optional[str] = None,
+        **options,
+    ):
+        """Draw fit line.
```
Addressed in 6bdc6ef.

conradhaupt commented on qiskit_experiments/visualization/style.py:
```diff
@@ -0,0 +1,122 @@
+        new.text_box_rel_pos: Tuple[float, float] = (0.6, 0.95)
+
+        # size of fit report text
+        new.text_box_text_size: int = 14
```
This style parameter is needed if we want `draw_text_box` in `BaseDrawer`. The comment was somewhat outdated as the parameter is not just for fit-reports, but I've fixed that.

conradhaupt commented on qiskit_experiments/curve_analysis/composite_curve_analysis.py:
```diff
@@ -310,23 +353,16 @@ def _run_analysis(
-                                        name=model._name + f"_{analysis.name}",
-                                        alpha=alpha,
-                                    )
+                                self.plotter.set_series_data(model._name, sigmas=sigmas)
```
Addressed in 9a909f5.

conradhaupt commented on qiskit_experiments/visualization/drawers/mpl_drawer.py:
```diff
@@ -0,0 +1,456 @@
+        """
+
+        def __init__(self, factor: float):
+            self.factor = factor
```
Addressed in 84ef573.

conradhaupt commented on qiskit_experiments/visualization/drawers/mpl_drawer.py:
```diff
@@ -0,0 +1,456 @@
+            self.factor = factor
+
+        def __call__(self, x, pos=None):
+            return self.fix_minus("{:.3g}".format(x * self.factor))
```
Addressed in 84ef573.

nkanazawa1989 reviewed (approved):
Thanks @conradhaupt for the awesome work! Now this PR looks good to me. Really excited to use new drawer (and much cool one with interactive backend).

eggerdj reviewed (approved):
LGTM

eggerdj reviewed (approved):
LGTM. Thanks for the naming update.
